### DISCUSSION TITLE
Save XYChart into a file

### DISCUSSION THREAD

**Discussion by asmirn1** [GSI-CS-CO/chart-fx#585] [Q&A]
Hi,

I'm generating a PDF file using the package `itextpdf` in Java. I'd like to add an `XYChart` image to that PDF file.

Is there a way to export `XYChart` into a JPEG or PNG file? So I could add that image to my PDF?

UPDATE:

I tried using the method `XYChart.snapshot()`, but it doesn't seem to export any image.

Here is the code that I run:
```java
package org.example;

import de.gsi.chart.XYChart;
import de.gsi.chart.axes.spi.DefaultNumericAxis;
import de.gsi.chart.renderer.spi.ErrorDataSetRenderer;
import de.gsi.dataset.spi.DefaultDataSet;
import javafx.application.Application;
import javafx.embed.swing.SwingFXUtils;
import javafx.scene.Scene;
import javafx.scene.SnapshotParameters;
import javafx.scene.image.WritableImage;
import javafx.stage.Stage;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.IOException;

public class Main extends Application {

    @Override
    public void start(Stage stage) throws IOException {

        // Create a chart
        XYChart chart = new XYChart(new DefaultNumericAxis("X"), new DefaultNumericAxis("Y"));
        ErrorDataSetRenderer renderer = new ErrorDataSetRenderer();
        DefaultDataSet dataSet = new DefaultDataSet("dataSet");
        dataSet.add(1, 2);
        dataSet.add(2, 3);
        dataSet.add(3, 4);
        renderer.getDatasets().add(dataSet);
        chart.getRenderers().add(renderer);

        // Save chart to a file
        SnapshotParameters parameters = new SnapshotParameters();
        WritableImage javaFxImage = chart.snapshot(parameters, null);
        BufferedImage swingImage = SwingFXUtils.fromFXImage(javaFxImage, null);
        ImageIO.write(swingImage, "png", new java.io.File("chart.png"));

        // Display the chart
        Scene scene = new Scene(chart, 640, 480);
        stage.setScene(scene);
        stage.show();
    }

    public static void main(String[] args) {
        launch();
    }
}
```

When I run this code, I can see the application window with the chart. Here is its screenshot:
<img width="752" alt="Screenshot 2023-05-03 at 2 09 05 PM" src="https://user-images.githubusercontent.com/13631120/236006069-5aa837bf-c8b1-49cd-8068-67ec2018323a.png">

However, when I look at the created file "chart.png", it's completely empty: [chart.png](https://user-images.githubusercontent.com/13631120/236006303-6c6f8a9d-a790-44ee-a28f-2deb4e727f3c.png)

Do you have any suggestions on what might be wrong?

This is my pom file if you need it:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>org.example</groupId>
    <artifactId>xychart_to_file</artifactId>
    <version>1.0-SNAPSHOT</version>

    <properties>
        <maven.compiler.source>20</maven.compiler.source>
        <maven.compiler.target>20</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    </properties>

    <dependencies>
        <dependency>
            <groupId>org.openjfx</groupId>
            <artifactId>javafx-controls</artifactId>
            <version>20</version>
        </dependency>
        <dependency>
            <groupId>org.openjfx</groupId>
            <artifactId>javafx-swing</artifactId>
            <version>20</version>
        </dependency>
        <dependency>
            <groupId>de.gsi.chart</groupId>
            <artifactId>chartfx-chart</artifactId>
            <version>11.2.7</version>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.openjfx</groupId>
                <artifactId>javafx-maven-plugin</artifactId>
                <version>0.0.8</version>
                <configuration>
                    <mainClass>org.example.Main</mainClass>
                </configuration>
            </plugin>
        </plugins>
    </build>

</project>
```

Alex

**Comment by wirew0rm** [comment]
you can look at how the screenshot plugin creates a snapshot:
https://github.com/ennerf/chart-fx/blob/main/chartfx-chart/src/main/java/io/fair_acc/chartfx/plugins/Screenshot.java#L230-L257

If you want to export the image non-interactively it might be necessary to enforce a few render passes before taking the snapshot due to some outstanding issues with scene invalidation.

**Comment by asmirn1** [comment]
Thank you for a quick answer! I'll take a look at the code.

**Comment by asmirn1** [comment]
@wirew0rm I've updated the question. You mentioned "a few render passes". How can I do that without actually displaying a chart?

**Reply by wirew0rm** [reply]
I suspect that you have to add the chart to the scene before taking the screenshot, otherwise it will not have a defined size.

**Reply by asmirn1** [reply]
@wirew0rm After moving `new Scene(chart, 640, 480)` in front of saving to a file, I've got some graphics:
 
![chart](https://user-images.githubusercontent.com/13631120/236256067-cfc4cb9b-f844-4170-b0e9-e3219182e053.png)

Now the axes and legend are displayed, but the plot (a line) and the grid are still missing. Do I need to do anything else?
